blockwork/build/caching.py

```diff
@@ -0,0 +1,159 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+'''
+This module contains the fundamentals required for the implementation of
+caching mechanisms. See the example project's cache for a simple working
+example.
+
+The cache scheme stores files using the hash of its content as a key, and
+additionally stores a many-to-one map of a hashkey which is computable
+before anything is run to that content hash.
+
+Stages:
+  Initial:
+    - Hash the content of static input interfaces
+    - Use these to compute transform hashkeys for nodes with no dependencies
+    - Use input interface names along with the hashkeys of transforms that 
+        output them to get the transform hashkeys for nodes with dependencies
+
+  Pre-run:
+    - Go through the transforms in reverse order
+    - Try and pull all output interfaces from caches - if successful mark that
+        transform as fetched.
+    - If all dependents of a transform are fetched, mark a transform as skipped
+
+  During-run:
+    - Go through transforms in dependency order as usual
+    - Skip transforms marked as fetchrd or skipped
+    - Push output interfaces to all caches that allow it
+
+Future improvements:
+  - Pass through information such as tags, file-size, and time-to-create 
+    through to caches so they can make intelligent decisions on what to cache.
+
+'''
+
+from abc import ABC, abstractmethod
+import hashlib
+import os
+from pathlib import Path
+from typing import Optional, TYPE_CHECKING
+if TYPE_CHECKING:
+    from blockwork.build.transform import Transform
+from ..context import Context
+
+# Switch to never pull or push from the cache, but to instead compare the 
+# hash with existing cache entries.
+# This will become a command option later...
+CACHE_CONSISTENCY_MODE = False
+
+class Cache(ABC):
+    @staticmethod
+    def hash_content(path: Path) -> str:
+        '''
+        Hash the content of a file or directory. This needs to be consistent
+        across caching schemes so consistency checks can be performed.
+        '''
+        if not path.exists():
+            raise RuntimeError
+        if path.is_dir():
+            content_hash = hashlib.md5('<dir>'.encode('utf8'))
+            for item in sorted(os.listdir(path)):
+                content_hash.update((item + Cache.hash_content(path / item)).encode('utf8'))
+        else:
+            with path.open('rb') as f:
+                content_hash = hashlib.file_digest(f, 'md5')
+        return content_hash.hexdigest()
+
+    @staticmethod
+    def store_all(ctx: Context, key_hash: str, frm: Path) -> bool:
```

Comment:
I was expecting `store_all` to store a list of stuff to a cache based on the name, perhaps `store_to_all`?

Reply:
Changed to store_to_any since this reflects the meaning of the return flag better (if any cache allowed it to be stored)